rfcs/0003-multi-blockchain-support.md

```diff
@@ -0,0 +1,534 @@
+  //   extend union DataSourceTemplate SubstrateDataSourceTemplate
+  //   ...
+  //
+  manifestExtensions: graphql.Document,
+  
+  // Type conversions to and from AssemblyScript.
+  // Conversions are defined as an array of tuples, each of which
+  // defines the source type (as a pattern or string), the target
+  // type (as a pattern or string), and a code snippet that performs
+  // the conversion.
+  //
+  // An example for Ethereum could look as follows:
+  typeConversions: {
+    fromAssemblyScript: [
+      ['Address', 'address', code => `EthereumValue.fromAddress(${code})`],
+      ['boolean', 'bool', code => `EthereumValue.fromBoolean(${code})`],
+      ['Bytes', 'byte', code => `EthereumValue.fromFixedBytes(${code})`],
+      ['Bytes', 'bytes', code => `EthereumValue.fromBytes(${code})`],
+      [
+        'Bytes',
+        /^bytes(1|2|3|4|5|6|7|8|9|10|11|12|13|14|15|16|17|18|19|20|21|22|23|24|25|26|27|28|29|30|31|32)$/,
+        code => `EthereumValue.fromFixedBytes(${code})`,
+      ],
+      ...
+    ],
+    toAssemblyScript: [
+      ['address', 'Address', code => `${code}.toAddress()`],
+      ['bool', 'boolean', code => `${code}.toBoolean()`],
+      ['byte', 'Bytes', code => `${code}.toBytes()`],
+      [
+        /^bytes(1|2|3|4|5|6|7|8|9|10|11|12|13|14|15|16|17|18|19|20|21|22|23|24|25|26|27|28|29|30|31|32)?$/,
+        'Bytes',
+        code => `${code}.toBytes()`,
+      ],
+      ...
+    ]
+  },
+  
+  // Code generation for ABIs,
+  codegen: (abi) => {
+    // Return a string to be put into the output file for the ABI
+  }
+}
+```
+
+Mapping files ending with `.ts` are automatically detected wherever they appear
+in the manifest and included in builds and deployments.
+
+#### Graph TS
+
+Each chain supported in mappings comes with its own `chains/<name>.ts` file in
+the `graph-ts` repository. Each of these has at least two sections: one for
+declaring host exports that are required for mappings to interact with the
+chain, another to define chain-specific types and helpers written in
+AssemblyScript. The file `chains/ethereum.ts` file could look as follows:
+
+```js
+import { Address } from '../index'
+
+export declare namespace ethereum {
+  function call(contract: Address, fn: string, args: Array<Value>): void
+}
+
+export namespace ethereum {
+    class Value { ... }
+    class Block { ... }
+    class Transaction { ... }
+    class Event { ... }
+    class Contract { ... }
+}
+```
+
+The main `index.ts` in the `graph-ts` repository re-exports each of these chain
+integrations via
+
+```js
+export * from './chains/ethereum'
+export * from './chains/substrate'
+```
+
+So that they can be imported into subgraph mappings with
+
+```js
+import { ethereum, substrate } from './chains/substrate'
+```
+
+## Compatibility
+
+The proposed design is mostly backwards-compatible, except for a few areas:
+
+- The manifest changes require a subgraph manifest migration in Graph CLI, as
```

Comment:
Is the main breaking change moving `abis` further up in the manifest? Could we support both patterns for some intermediate amount of time so that all subgraphs manifests don't need to be updated immediately?

Reply:
So far what we've done is automatically migrate manifests at `graph codegen/build` time and that's worked well.